src/Lean/Widget/ExprWithCtx.lean

```diff
@@ -0,0 +1,121 @@
+/-
+Copyright (c) 2021 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+
+Authors: Wojciech Nawrocki
+-/
+import Lean.PrettyPrinter
+import Lean.Server.Rpc.Basic
+import Lean.Server.Rpc.RequestHandling
+import Lean.Widget.TaggedText
+
+/-! RPC infrastructure for storing/formatting `Expr`s with environment and subexpression information. -/
+
+namespace Lean.Widget
+open Server
+
+/-- Bundles an `Expr` with its `MetaM` context. -/
+structure ExprWithCtx where
+  expr          : Expr
+  env           : Environment
+  mctx          : MetavarContext
+  lctx          : LocalContext
+  options       : Options
+  currNamespace : Name
+  openDecls     : List OpenDecl
+  deriving Inhabited, RpcEncoding with { withRef := true }
+
+/-- We traverse expressions lazily when the client requests it. -/
+abbrev LazyExprWithCtx := Unit → IO ExprWithCtx
```

Comment:
Thunks are pure, right? So it'd have to be `Thunk (IO ExprWithCtx)` which I guess caches the action but not its result, which isn't very useful. I'm not aware of a `ThunkIO`. `Task` would work but it's kind of weird to thread-pool this. As for using `LazyExprWithCtx` everywhere, I guess that's reasonable enough. I'll try it.

Reply:
There's `unsafeIO`, so you can write a very reasonable `IO α → IO (Thunk (Except IO.Error α))` function.

Reply:
I did not carefully look at this PR yet, but please try to avoid `unsafeIO` as much as possible. We had many issues with it in the past, the current implementation uses it in a single place: `profileitIO`, and we should probably remove it in the future. In the past, we used `unsafeIO` in other places, and if I remember correctly it "blew up" in our faces more than once. It is easy to write incorrect code when using `unsafeIO`. BTW, if I remember correctly, we had problems with a very similar function
```
constant asTask (act : IO α) (prio := Task.Priority.default) : IO (Task (Except IO.Error α))
```
Note that it is not marked as `unsafe`, but `act` is always executed. I think the same justification does not work for `IO α → IO (Thunk (Except IO.Error α))` since it would not make sense to run the IO action eagerly in this case. So, I am assuming the function `IO α → IO (Thunk (Except IO.Error α))` would have to be marked as `unsafe`.
The following link has a discussion of a similar issue in Haskell. https://mail.haskell.org/pipermail/haskell/2009-March/021064.html
@Kha Do you remember the details? I recall discussing these issues with you but I forgot the details.

Reply:
This is interesting, I didn't realize that nondeterministic execution fell under the unsafe umbrella as well.

I still think it would still be a good idea to avoid the double references (plus the extra server roundtrip for the `LazyExprWithCtx.get` call).  A `ThunkIO` implementation that requires `IO` to access the thunk should still be fine, right?
```lean
def ThunkIO (α) := IO.Ref (Sum (IO α) α)

def ThunkIO.mk (a : IO α) : IO (ThunkIO α) := IO.mkRef (Sum.inl a)

def ThunkIO.get (t : ThunkIO α) : IO α := do
  match ← ST.Ref.get t with
    | Sum.inr a => a
    | Sum.inl fn =>
      let a ← fn -- potential double evaluation, but avoids locking
      t.set (Sum.inl a)
      a
```

Reply:
Now that we are using `Info`s which are produced by the delaborator eagerly, `LazyExprWithCtx` is no longer necessary. For lazy traces we use `MsgEmbed.lazyTrace`.